### PR TITLE
Refactor handling of coefficients so they are trackable

### PR DESCRIPTION
**Pull request by cmacmackin**
These changes mean that Coefficient objects can be created which are tracked by an `ObjectManager` class. This means they can all easily be updated (e.g., with the latest time-step). This also allows the convenience of automatically creating MFEM coefficients whenever a MOOSE function is declared, so it can be reused.

### CODE REVIEW

File: src/coefficients/MFEMConstantCoefficient.C [hunk 1]
```diff
@@ -12,7 +13,8 @@ MFEMConstantCoefficient::validParams()
 
 MFEMConstantCoefficient::MFEMConstantCoefficient(const InputParameters & parameters)
   : MFEMCoefficient(parameters),
-    coefficient{std::make_shared<mfem::ConstantCoefficient>(getParam<double>("value"))}
+    coefficient{getMFEMProblem().getProblemData()._scalar_manager.make<mfem::ConstantCoefficient>(
+        getParam<double>("value"))}
```
Comment (alexanderianblair) [inline src/coefficients/MFEMConstantCoefficient.C:17]: Would it be worth adding a `addScalarCoefficient<T>(args...)` method to `MFEMProblem` that wraps the `getProblemData()._scalar_manager.make<T>(args...)` call for readability?

I understand we're expecting to remove the `MFEMCoefficient` classes as objects that end users can create from the input stack directly, but I suspect developers may still want to create (tracked) coefficients in some kernels/BCs/Executioners, where such a method on `MFEMProblem` would help.

(Similarly for vector and matrix coefficients)

File: test/tests/kernels/curlcurl.i [hunk 2]
```diff
@@ -62,10 +62,6 @@
     type = MFEMVectorFunctionCoefficient
     function = exact_e_field
```
Comment (alexanderianblair) [inline test/tests/kernels/curlcurl.i:63]: Function Dirichlet BCs can also now take the function name directly, now `addFunction` automatically sets the coefficient (but assuming that's planned for a later PR when `MFEMCoefficients` are removed?)

**Review by alexanderianblair — COMMENTED** [review]
Looks good - allows function specification in BCs/kernels inputs to be much more consistent with MOOSE, whilst improving `PropertyManager`'s handling of Coefficients.

Couple of suggestions/questions, but nothing blocking other than a rebase needed

File: test/tests/kernels/curlcurl.i (hunk 2 above)
Comment (cmacmackin) [reply test/tests/kernels/curlcurl.i:63]: Yes, I'm leaving that for the next PR (which I'll probably open tomorrow).

File: src/coefficients/MFEMConstantCoefficient.C (hunk 1 above)
Comment (cmacmackin) [reply src/coefficients/MFEMConstantCoefficient.C]: I've created those convenience methods. They will be useful for the new boundary condition classes in the next PR.

**Comment by alexanderianblair** [comment]
Hotfix merged - and yes, assuming the tests all pass, this is good to merge

**Comment by cmacmackin** [comment]
@alexanderianblair All ready to merge now.

**Review by alexanderianblair — APPROVED** [review]
